src/order/grade.lean

```diff
@@ -0,0 +1,334 @@
+element by how many elements we need to go down to reach `⊥`. Unpractical, really.
+
+Instead, we define graded orders by their grade function, without talking about flags yet.
+
+## References
+
+* [Konrad Engel, *Sperner Theory*][engel1997]
+* [Richard Stanley, *Enumerative Combinatorics*][stanley2012]
+-/
+
+open finset nat order_dual
+
+variables {α β : Type*}
+
+/-- A graded order is an order equipped with a grade function which tells how far a given element is
+away from the minimal elements. Precisely, `grade a` is the height of `a` in the Hasse diagram of
+`α`. -/
+class grade_order (α : Type*) [preorder α] :=
+(grade : α → ℕ)
+(grade_strict_mono : strict_mono grade)
+(grade_of_is_min ⦃a : α⦄ : is_min a → grade a = 0)
+(grade_of_covers ⦃a b : α⦄ : a ⋖ b → grade a + 1 = grade b)
+
+/-- A graded top order is a graded order with a maximal grade (NOT a maximal *element*). -/
+class grade_max_order (α : Type*) [preorder α] extends grade_order α :=
+(max_grade : ℕ)
+(grade_le_max_grade (a : α) : grade a ≤ max_grade)
+(grade_of_is_max ⦃a : α⦄ : is_max a → grade a = max_grade)
+
+section grade_order
+section preorder
+variables [preorder α] [grade_order α] {a b : α}
+
+/-- The grade of an element in a graded order. Morally, this is the number of elements you need to
+go down by to get to `⊥`. -/
+def grade (a : α) : ℕ := grade_order.grade a
+lemma grade_strict_mono : strict_mono (grade : α → ℕ) := grade_order.grade_strict_mono
+
+protected lemma is_min.grade (h : is_min a) : grade a = 0 := grade_order.grade_of_is_min h
+protected lemma covers.grade (h : a ⋖ b) : grade a + 1 = grade b := grade_order.grade_of_covers h
+
+/-- If two elements in a graded partial order cover each other, so do their grades. This is just a
+restatement of the covering condition. -/
+lemma covers.grade_covers (h : a ⋖ b) : grade a ⋖ grade b := covers_iff_succ_eq.2 h.grade
+
+lemma covers_iff_grade_succ_eq_lt : a ⋖ b ↔ grade a + 1 = grade b ∧ a < b :=
+⟨λ h, ⟨h.grade, h.1⟩, λ h, ⟨h.2, λ c ha hb,
+  (covers_iff_succ_eq.2 h.1).2 (grade_strict_mono ha) $ grade_strict_mono hb⟩⟩
+
+@[simp] lemma grade_eq_zero_iff : grade a = 0 ↔ is_min a :=
+begin
+  refine ⟨λ h b hba, _, is_min.grade⟩,
+  by_contra hab,
+  exact not_lt_bot ((grade_strict_mono $ hba.lt_of_not_le hab).trans_le h.le),
+end
+
+section order_bot
+variables [order_bot α]
+
+@[simp] lemma grade_bot : grade (⊥ : α) = 0 := is_min_bot.grade
+
+end order_bot
+
+section order_top
+variables [order_top α]
+
+lemma has_lt.lt.grade_lt_grade_top (h : a < b) : grade a < grade (⊤ : α) :=
+grade_strict_mono $ h.trans_le le_top
+
+@[simp] lemma grade_lt_grade_top_of_nonempty_Ioi (h : (set.Ioi a).nonempty) :
+  grade a < grade (⊤ : α) :=
+has_lt.lt.grade_lt_grade_top h.some_mem
+
+end order_top
+end preorder
+
+section partial_order
+variables [partial_order α] [grade_order α] {a b : α}
+
+lemma grade_mono : monotone (grade : α → ℕ) := grade_strict_mono.monotone
```

Comment:
Why do we need the partial order for this? Can't we directly show it's monotone from the fact that it's strict mono without this extra assumption?

Reply:
No, this is a well-known thing about strict monotonicity (cf the docstring of `strict_mono.monotone`. It holds vacuously when `≤` is always true.